### PR TITLE
advertiser and data provider support

### PR DESCRIPTION
**Pull request by jon8787**

### CODE REVIEW

**Comment by sunnywu** [comment]
Could you please add descriptions on the readme.md  for the IdentityMapClient, similar to PublisherClient description? https://github.com/IABTechLab/uid2-client-java#usage-for-publishers

And also copy these new description onto uid2docs on this page?
https://unifiedid.com/docs/sdks/uid2-sdk-ref-java

File: src/main/java/com/uid2/client/InputUtil.java [hunk 1]
```diff
@@ -33,6 +33,15 @@ public static boolean isPhoneNumberNormalized(String phoneNumber) {
     return totalDigits >= MIN_PHONENUMBER_DIGITS && totalDigits <= MAX_PHONENUMBER_DIGITS;
   }
 
+  public static String normalizeAndHashEmail(String unnormalizedEmail) {
```
[the remark this answers is not in the document]
Comment (jon8787) [reply src/main/java/com/uid2/client/InputUtil.java:36]: done :)

File: src/main/java/com/uid2/client/IdentityMapInput.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    public static IdentityMapInput fromHashedPhones(Iterable<String> hashedPhones) {
+        return new IdentityMapInput(IdentityType.Phone, hashedPhones, true);
+    }
+
+    private IdentityMapInput(IdentityType identityType, Iterable<String> emailsOrPhones, boolean alreadyHashed) {
+        if (identityType == IdentityType.Email) {
+            hashedNormalizedEmails = new ArrayList<>();
+            for (String email : emailsOrPhones) {
+                if (alreadyHashed) {
+                    hashedNormalizedEmails.add(email);
```
Comment (atarassov-ttd) [inline src/main/java/com/uid2/client/IdentityMapInput.java:46]: I think the answer is yet, but does operator return hash exactly as passed without transformations? E.g. lowercase, etc. What happens if the same hash is specified multiple times on input?

File: src/main/java/com/uid2/client/IdentityMapInput.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                } else {
+                    String hashedEmail = InputUtil.normalizeAndHashEmail(email);
```
Comment (atarassov-ttd) [inline src/main/java/com/uid2/client/IdentityMapInput.java:48]: What happens if input contains multiple instances of the same email or variants of it that result in the same hash?

File: src/main/java/com/uid2/client/IdentityMapInput.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * @param hashedPhones a <a href="https://unifiedid.com/docs/getting-started/gs-normalization-encoding#phone-number-normalization">normalized</a> and <a href="https://unifiedid.com/docs/getting-started/gs-normalization-encoding#phone-number-hash-encoding">hashed</a> phone number
+     * @return an IdentityMapInput instance
+     */
+    public static IdentityMapInput fromHashedPhones(Iterable<String> hashedPhones) {
+        return new IdentityMapInput(IdentityType.Phone, hashedPhones, true);
+    }
+
+    private IdentityMapInput(IdentityType identityType, Iterable<String> emailsOrPhones, boolean alreadyHashed) {
```
Comment (atarassov-ttd) [inline src/main/java/com/uid2/client/IdentityMapInput.java:41]: What happens if `emailsOrPhones` is empty?

File: src/main/java/com/uid2/client/IdentityMapResponse.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.annotations.SerializedName;
+
+import java.util.HashMap;
+
+public class IdentityMapResponse {
+    IdentityMapResponse(String response, IdentityMapInput identityMapInput) {
```
Comment (atarassov-ttd) [inline src/main/java/com/uid2/client/IdentityMapResponse.java:12]: nit: Is it worth turning this into a factory function?

Comment (jon8787) [reply src/main/java/com/uid2/client/IdentityMapResponse.java:12]: I don't anticipate implementation differences between UID2 & EUID that wouldn't also require an interface change, and that couldn't be handled server-side.
The one place where we did use factories was Sharing encryption - but even there I think returning whether we're EUID vs UID2 from the server would've been better.

File: src/main/java/com/uid2/client/IdentityMapInput.java (hunk 2 above)
Comment (jon8787) [reply src/main/java/com/uid2/client/IdentityMapInput.java:46]: Yes, the hash is base64 encoded so it's case sensitive and is returned in the same case. 
Same hash multiple times is deduplicated in the response. I've added a test `identityMapDuplicateHashedEmails`

File: src/main/java/com/uid2/client/IdentityMapInput.java (hunk 4 above)
Comment (jon8787) [reply src/main/java/com/uid2/client/IdentityMapInput.java:41]: The request still works but getMappedIdentities() and getUnmappedIdentities return empty maps. Added test `identityMapEmptyInput`

File: src/main/java/com/uid2/client/IdentityMapInput.java (hunk 3 above)
Comment (jon8787) [reply src/main/java/com/uid2/client/IdentityMapInput.java:49]: thanks, good spot. Fixed and added `identityMapDuplicateEmails` test

File: src/main/java/com/uid2/client/IdentityMapResponse.java (hunk 5 above)
Comment (jon8787) [reply src/main/java/com/uid2/client/IdentityMapResponse.java:13]: Oh, the ctor is package-private. SDK consumers can't create these other than by calling IdentityMapClient.generateIdentityMap or IdentityMapHelper.createIdentityMapResponse

**Pull request merged**